test/demos/forwarddiffzero.jl

```diff
@@ -0,0 +1,76 @@
+"The simplest viable forward mode a AD, only supports `Float64`"
+module ForwardDiffZero
+
+using ChainRulesCore
+using Test
+
+struct Dual <: Real
+    primal::Float64
+    diff::Float64
+end
+
+primal(d::Dual) = d.primal
+diff(d::Dual) = d.diff
+
+primal(d::Real) = d
+diff(d::Real) = 0.0
+
+# needed for ^ to work from having `*` defined
+Base.to_power_type(x::Dual) = x
+
+
+function define_dual_overload(sig, ast)
+    opT, argTs = Iterators.peel(sig.parameters)
+    fieldcount(opT) == 0 || return  # not handling functors 
+    all(Float64 <: argT for argT in argTs) || return  # only handling purely Float64 ops.
+    op = opT.instance
+    opname = :($(parentmodule(op)).$(nameof(op)))
+    N = length(sig.parameters) - 1  # skip the op
+    fdef = quote
+        function $opname(dual_args::Vararg{Union{Dual, Float64},$N}; kwargs...)
+            ȧrgs = (NO_FIELDS,  diff.(dual_args)...)
+            args = ($opname, primal.(dual_args)...)
+            res = frule(ȧrgs, args...; kwargs...)
+            res === nothing && error("Apparently no rule for $($sig)), but we really thought there was, args=($args)")
+            y, ẏ = res
+            return Dual(y, ẏ)  # if y, ẏ are not `Float64` this will error.
+        end
+    end
+    # @show fdef
+    @eval $fdef
+end
```

Comment:
because this code never uses the `AST` I am now questioning if we even need to capture that.
If we don't then that is great because we can get rid of the `FRULES` and `RRULES`,
and just use `methods(frule)` and `methods(rrule)` instead.

Reply:
Wasn't the rationale here that a Zygote-like system would benefit from being able to manipulate the AST or something? Maybe I'm misremembering.

Reply:
We maybe could even get rid of the `@frule` and `@rrule` macros then, and go back to original proposal of using hooks attached to `on_package_load`

Reply:
> using hooks attached to `on_package_load`

What is this proposal? Is it written down somewhere? If the hooks are not triggered by `@frule`/`@rrule` macros, how are they triggered? Some Revise-style magic?

Reply:
It is mentioned here
https://github.com/JuliaDiff/ChainRulesCore.jl/issues/127#issuecomment-599216325
and yes it is the same hook in `Base` that power's Revise:
`Base.package_callbacks`

it would be triggerd whenever a package is loaded.
Probably we would provide a manual `ChainRulesCore.refresh` also for people to use in the REPL

Reply:
If `@eval` will be called by the AD package, then we run into namespace issues. The correct behavior for dispatch-based AD is to overload the AD package's methods in the correct namespace outside the AD package, i.e. in the module where the rule is defined. Is this possible using this hook mechanism?

Reply:
Yes that is what is done in the ForwardDiffZero demo.
It does:
```
@eval function Base.+(::Dual, ::Dual)
...
end
```

Reply:
But what if `Base` is replaced with another module and `...` uses code from that other module, and this module is not available in ForwardDiffZero?

Reply:
`@eval` here evaluates the code in `ForwardDiffZero` so it will complain if we use code not loaded in ForwardDiffZero iiuc.

Reply:
> But what if Base is replaced with another module and ... uses code from that other module, and this module is not available in ForwardDiffZero?

the `...` is being written by the author of `ForwardDiffZero` and i don't think has any reason to run anything other than functions from `ForwardDiffZero`, or from `ChainRuleCore` (which is loaded by `ForwardDiffZero`).
Probably this is a reason not to include the AST as trying to use that for something in `...` will run into this.
I think its just not needed anyway.

The interesting bit is I guess is in the sig, for a Source to Source AD, like Zygote.
that is not generating overloads of `op(overloaded_equiv.(args)....)`, but rather `pullback(op, args...)`
Maybe the types of the args would need the same `  opname = :($(parentmodule(op)).$(nameof(op)))` type escaping,
though even that wouldn't work because it would be a `Symbol`.
But can it be a type directly and thus not need to be given a path in local scope?
Yes, that seems to work
```julia
julia> K = Base.Fix2
Base.Fix2

julia> eval(:(foo(x::$K) = x))
foo (generic function with 1 method)

julia> foo(Base.Fix2(+,1))
(::Base.Fix2{typeof(+),Int64}) (generic function with 1 method)
```

Reply:
We can also do this trick to avoid qualifying names for the operation.
Though it has to be in call overload form for some reason.
(Still i guess that is fine since more generic if also overloading functors)


```julia
julia> struct Foo end

julia> K = typeof(Base.:+)  # this would come in the sig tuple, just using Base as example
typeof(+)

julia> @eval (::$K)(::Foo, ::Foo) = 2

julia> +(Foo(), Foo())
2

julia> +(2,1)
3

julia> methods(+, (Foo, Foo))
# 1 method for generic function "+":
[1] +(::Foo, ::Foo) in Main at REPL[17]:1

julia> methods(+)
# 167 methods for generic function "+":
[1] +(x::Bool, z::Complex{Bool}) in Base at complex.jl:282
[2] +(x::Bool, y::Bool) in Base at bool.jl:96
[3] +(x::Bool) in Base at bool.jl:93
```

Reply:
I am now using this trick in both demos.
Can this be considered resolved?